File: p010.c
/*
** summation of primes
*/

#include <stdio.h>
#include <stdlib.h>

#define LIMIT 2000000

int prime(int n, int *p, int m) {
    for (int k = 0; k < m; k++) {
        if (n % p[k] == 0) return 0;
        if (p[k] * p[k] >= n) return 1;
    }
    return 1;
}

int main(void) {
    int *p = calloc(2000000, sizeof *p);
    int np = 1;
    *p = 2;
    int n = 3;
    long long unsigned sum = 2;
    while (n < LIMIT) {
        if (prime(n, p, np)) {
            // add n to p
            p[np++] = n;
            sum += n;
        }
        n += 2;
    }
    printf("sum is %llu\n", sum);
}

File: forth/p010.fs
\ Summation of primes

\ Problem 10

\ The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.

\ Find the sum of all the primes below two million.

\ no need to check for 2 or even numbers
: e007-prime?
    3 ( n factor ) begin 2dup dup * >= while
                   2dup mod 0= if 2drop false exit then 2 + repeat
    2drop true ;

: e010-sumprimes
    2 swap 3 ( sum limit curr )
    begin 2dup >= while
        dup e007-prime? if rot over + rot rot then 2 +
    repeat drop drop ;

: euler010
    2000000 e010-sumprimes . ;
